bin/reth-bench/src/valid_payload.rs

```diff
@@ -0,0 +1,177 @@
+
+use alloy_provider::{ext::EngineApi, Network};
+use alloy_rpc_types_engine::{
+    ExecutionPayloadInputV2, ForkchoiceState, ForkchoiceUpdated, PayloadAttributes, PayloadStatus,
+    PayloadStatusEnum,
+};
+use alloy_transport::{Transport, TransportResult};
+use reth_primitives::B256;
+use reth_rpc_types::{ExecutionPayloadV1, ExecutionPayloadV3};
+use tracing::error;
+
+/// An extension trait for providers that implement the engine API, to wait for a VALID response.
+#[async_trait::async_trait]
+pub trait EngineApiValidWaitExt<N, T>: Send + Sync {
+    /// Calls `engine_newPayloadV1` with the given [ExecutionPayloadV1], and waits until the
+    /// response is VALID.
+    async fn new_payload_v1_wait(
+        &self,
+        payload: ExecutionPayloadV1,
+    ) -> TransportResult<PayloadStatus>;
+
+    /// Calls `engine_newPayloadV2` with the given [ExecutionPayloadInputV2], and waits until the
+    /// response is VALID.
+    async fn new_payload_v2_wait(
+        &self,
+        payload: ExecutionPayloadInputV2,
+    ) -> TransportResult<PayloadStatus>;
+
+    /// Calls `engine_newPayloadV3` with the given [ExecutionPayloadV3], parent beacon block root,
+    /// and versioned hashes, and waits until the response is VALID.
+    async fn new_payload_v3_wait(
+        &self,
+        payload: ExecutionPayloadV3,
+        versioned_hashes: Vec<B256>,
+        parent_beacon_block_root: B256,
+    ) -> TransportResult<PayloadStatus>;
+
+    /// Calls `engine_forkChoiceUpdatedV1` with the given [ForkchoiceState] and optional
+    /// [PayloadAttributes], and waits until the response is VALID.
+    async fn fork_choice_updated_v1_wait(
+        &self,
+        fork_choice_state: ForkchoiceState,
+        payload_attributes: Option<PayloadAttributes>,
+    ) -> TransportResult<ForkchoiceUpdated>;
+
+    /// Calls `engine_forkChoiceUpdatedV2` with the given [ForkchoiceState] and optional
+    /// [PayloadAttributes], and waits until the response is VALID.
+    async fn fork_choice_updated_v2_wait(
+        &self,
+        fork_choice_state: ForkchoiceState,
+        payload_attributes: Option<PayloadAttributes>,
+    ) -> TransportResult<ForkchoiceUpdated>;
+
+    /// Calls `engine_forkChoiceUpdatedV3` with the given [ForkchoiceState] and optional
+    /// [PayloadAttributes], and waits until the response is VALID.
+    async fn fork_choice_updated_v3_wait(
+        &self,
+        fork_choice_state: ForkchoiceState,
+        payload_attributes: Option<PayloadAttributes>,
+    ) -> TransportResult<ForkchoiceUpdated>;
+}
+
+#[async_trait::async_trait]
+impl<T, N, P> EngineApiValidWaitExt<N, T> for P
+where
+    N: Network,
+    T: Transport + Clone,
+    P: EngineApi<N, T>,
+{
+    async fn new_payload_v1_wait(
+        &self,
+        payload: ExecutionPayloadV1,
+    ) -> TransportResult<PayloadStatus> {
+        // TODO: remove clones somehow?
+        let mut status = self.new_payload_v1(payload.clone()).await?;
+        // TODO: log invalids
```

Comment:
log and should probably return invalids right? same goes for accepted, no? why call again in a loop with same payload expecting a different result?

Reply:
looping when syncing may be ok, some log message should defo be printed then

Reply:
error logging + panicking on invalid now, I think logging on SYNCING would be a bit too noisy, since it retries immediately